src/edu/washington/escience/myria/expression/evaluate/PythonUDFEvaluator.java

```diff
@@ -60,21 +52,20 @@
       org.slf4j.LoggerFactory.getLogger(PythonUDFEvaluator.class);
   /** python function registrar from which to fetch function pickle. */
   private final PythonFunctionRegistrar pyFuncRegistrar;
-
   /** python worker process. */
   private PythonWorker pyWorker;
   /** index of state column. */
-  private final boolean[] isStateColumn;
-  /** tuple size to be sent to the python process, this is equal to the number of children of the expression. */
-  private int numColumns = -1;
-
+  private Set<Integer> stateColumns;
+  /** column indices of child ops. */
   private int[] columnIdxs = null;
-
   /** Output Type of the expression. */
   private Type outputType = null;
-
   /** is expression a flatmap? */
   private Boolean isMultiValued = false;
+  /** Tuple buffers for each group key. */
+  private IntObjectHashMap<TupleBatchBuffer> buffer;
```

Comment:
It's a little bit awkward, since each entry in this `buffer` was designed to match a row of the `state` in `Aggregate`.  If we change it to a `TupleHashTable`, then we need to pass the values of keys of each row in `state` to this `TupleHashTable`, hash them, and check if the keys indeed match, in order to send all tuples of a group to Python. I think there is some CPU overhead. One the other hand, the memory overhead of this design comparing to using `TupleHashTable`, is the Java object overhead of a list of Columns per group key, which may not be very bad to me. I agree that it would be the best to get rid of this memory overhead without sacrificing time complexity.